[V2] ControllerUnpublishVolume panics due to nil object in type conversion

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!
-->


**What happened**:

Shortly after the V2 driver started it received a `ControllerUnpublishVolume` call and failed an interface conversion due to a `nil` object.

```log
I0811 22:18:46.779850       1 utils.go:78] GRPC call: /csi.v1.Controller/ControllerUnpublishVolume
I0811 22:18:46.780138       1 utils.go:79] GRPC request: {"node_id":"capz-noxtnr-md-0-g5bjg","volume_id":"/subscriptions/0e46bd28-a80f-4d3a-8200-d9eb8d80cb2e/resourceGroups/capz-noxtnr/providers/Microsoft.Compute/disks/pvc-d796e0dd-b7ff-4a07-bd96-ce252d58ea19"}
I0811 22:18:46.780988       1 crdprovisioner.go:922]  "msg"="Requesting AzVolumeAttachment (pvc-d796e0dd-b7ff-4a07-bd96-ce252d58ea19-capz-noxtnr-md-0-g5bjg-attachment) detachment" "disk.csi.azure.com/node-name"="capz-noxtnr-md-0-g5bjg" "disk.csi.azure.com/request-id"="da685217-19c1-11ed-ae9c-000d3ad6933f" "disk.csi.azure.com/requested-role"="Primary" "disk.csi.azure.com/volume-name"="pvc-d796e0dd-b7ff-4a07-bd96-ce252d58ea19"
I0811 22:18:46.781364       1 azure_disk_utils.go:1015]  "msg"="Workflow completed with success." "caller"="sigs.k8s.io/azuredisk-csi-driver/pkg/provisioner.(*CrdProvisioner).detachVolume" "disk.csi.azure.com/node-name"="capz-noxtnr-md-0-g5bjg" "disk.csi.azure.com/request-id"="da685217-19c1-11ed-ae9c-000d3ad6933f" "disk.csi.azure.com/requested-role"="Primary" "disk.csi.azure.com/requester-name"="sigs.k8s.io/azuredisk-csi-driver/pkg/azureutils.UpdateCRIWithRetry" "disk.csi.azure.com/volume-name"="pvc-d796e0dd-b7ff-4a07-bd96-ce252d58ea19" "latency"=197847
I0811 22:18:46.781542       1 crdprovisioner.go:833]  "msg"="Workflow completed with success." "disk.csi.azure.com/node-name"="capz-noxtnr-md-0-g5bjg" "disk.csi.azure.com/request-id"="917184ed-19c3-11ed-8be1-000d3ad6933f" "disk.csi.azure.com/requested-role"="Primary" "disk.csi.azure.com/requester-name"="sigs.k8s.io/azuredisk-csi-driver/pkg/provisioner.(*CrdProvisioner).UnpublishVolume" "disk.csi.azure.com/volume-name"="pvc-d796e0dd-b7ff-4a07-bd96-ce252d58ea19" "latency"=817392
I0811 22:18:46.781773       1 azure_metrics.go:114] "Observed Request Latency" latency_seconds=0.001144869 request="azuredisk_csi_driver_controller_unpublish_volume" resource_group="capz-noxtnr" subscription_id="0e46bd28-a80f-4d3a-8200-d9eb8d80cb2e" source="disk.csi.azure.com" volumeid="/subscriptions/0e46bd28-a80f-4d3a-8200-d9eb8d80cb2e/resourceGroups/capz-noxtnr/providers/Microsoft.Compute/disks/pvc-d796e0dd-b7ff-4a07-bd96-ce252d58ea19" node="capz-noxtnr-md-0-g5bjg" result_code="failed"
panic: interface conversion: client.Object is nil, not *v1beta2.AzVolumeAttachment

goroutine 240 [running]:
sigs.k8s.io/azuredisk-csi-driver/pkg/provisioner.(*CrdProvisioner).detachVolume(0xc0001d04d0, {0x208d238, 0xc00058bfb0}, 0xc00063a9c0)
	/home/prow/go/src/sigs.k8s.io/azuredisk-csi-driver/pkg/provisioner/crdprovisioner.go:927 +0x8b4
sigs.k8s.io/azuredisk-csi-driver/pkg/provisioner.(*CrdProvisioner).UnpublishVolume(0xc0001d04d0, {0x208d238, 0xc00058af90}, {0xc00025a0a0?, 0x1dc6aa9?}, {0xc00014b938, 0x16}, 0xb?, 0xc00058da70?)
	/home/prow/go/src/sigs.k8s.io/azuredisk-csi-driver/pkg/provisioner/crdprovisioner.go:851 +0xa5c
sigs.k8s.io/azuredisk-csi-driver/pkg/azuredisk.(*DriverV2).ControllerUnpublishVolume(0xc0000af680, {0x208d238, 0xc00058af90}, 0xc000408320)
	/home/prow/go/src/sigs.k8s.io/azuredisk-csi-driver/pkg/azuredisk/controllerserver_v2.go:285 +0x36b
github.com/container-storage-interface/spec/lib/go/csi._Controller_ControllerUnpublishVolume_Handler.func1({0x208d238, 0xc00058af90}, {0x1c90b00?, 0xc000408320})
	/home/prow/go/src/sigs.k8s.io/azuredisk-csi-driver/vendor/github.com/container-storage-interface/spec/lib/go/csi/csi.pb.go:5730 +0x78
sigs.k8s.io/azuredisk-csi-driver/pkg/csi-common.logGRPC({0x208d238, 0xc00058af90}, {0x1c90b00?, 0xc000408320?}, 0xc000806880, 0xc0007c4d68)
	/home/prow/go/src/sigs.k8s.io/azuredisk-csi-driver/pkg/csi-common/utils.go:81 +0x3a3
github.com/container-storage-interface/spec/lib/go/csi._Controller_ControllerUnpublishVolume_Handler({0x1d7db00?, 0xc0000af680}, {0x208d238, 0xc00058af90}, 0xc00052c4e0, 0x1ea9a00)
	/home/prow/go/src/sigs.k8s.io/azuredisk-csi-driver/vendor/github.com/container-storage-interface/spec/lib/go/csi/csi.pb.go:5732 +0x138
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000290e00, {0x2093fb8, 0xc00088b980}, 0xc000254b40, 0xc00024a450, 0x2f178a8, 0x0)
	/home/prow/go/src/sigs.k8s.io/azuredisk-csi-driver/vendor/google.golang.org/grpc/server.go:1282 +0xccf
google.golang.org/grpc.(*Server).handleStream(0xc000290e00, {0x2093fb8, 0xc00088b980}, 0xc000254b40, 0x0)
	/home/prow/go/src/sigs.k8s.io/azuredisk-csi-driver/vendor/google.golang.org/grpc/server.go:1616 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	/home/prow/go/src/sigs.k8s.io/azuredisk-csi-driver/vendor/google.golang.org/grpc/server.go:921 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/home/prow/go/src/sigs.k8s.io/azuredisk-csi-driver/vendor/google.golang.org/grpc/server.go:919 +0x28a
```

**What you expected to happen**:

The `ControllerUnpublishVolume` call should not panic.

**How to reproduce it**:

**Anything else we need to know?**:

`UpdateCRIWithRetry` appears to have returned a `nil` updated object, likely because no update was *actually* made.

**Environment**:
- CSI Driver version:
- Kubernetes version (use `kubectl version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

--- a/pkg/azureutils/azure_disk_utils.go
+++ b/pkg/azureutils/azure_disk_utils.go
@@ -1018,45 +1018,32 @@ func UpdateCRIWithRetry(ctx context.Context, informerFactory azdiskinformers.Sha
 
 	conditionFunc := func() error {
 		var err error
-		var copyForUpdate client.Object
-		var objForUpdate client.Object
+		var originalObj client.Object
 		switch target := obj.(type) {
 		case *azdiskv1beta2.AzVolume:
-			updateTarget := &azdiskv1beta2.AzVolume{}
 			if informerFactory != nil {
-				updateTarget, err = informerFactory.Disk().V1beta2().AzVolumes().Lister().AzVolumes(target.Namespace).Get(objName)
+				originalObj, err = informerFactory.Disk().V1beta2().AzVolumes().Lister().AzVolumes(target.Namespace).Get(objName)
 			} else if cachedClient != nil {
-				err = cachedClient.Get(ctx, types.NamespacedName{Namespace: target.Namespace, Name: objName}, updateTarget)
+				originalObj = &azdiskv1beta2.AzVolume{}
+				err = cachedClient.Get(ctx, types.NamespacedName{Namespace: target.Namespace, Name: objName}, originalObj)
 			} else {
-				updateTarget, err = azDiskClient.DiskV1beta2().AzVolumes(target.Namespace).Get(ctx, objName, metav1.GetOptions{})
-			}
-			if err == nil {
-				objForUpdate = updateTarget
-				copyForUpdate = updateTarget.DeepCopy()
+				originalObj, err = azDiskClient.DiskV1beta2().AzVolumes(target.Namespace).Get(ctx, objName, metav1.GetOptions{})
 			}
 		case *azdiskv1beta2.AzVolumeAttachment:
-			updateTarget := &azdiskv1beta2.AzVolumeAttachment{}
 			if informerFactory != nil {
-				updateTarget, err = informerFactory.Disk().V1beta2().AzVolumeAttachments().Lister().AzVolumeAttachments(target.Namespace).Get(objName)
+				originalObj, err = informerFactory.Disk().V1beta2().AzVolumeAttachments().Lister().AzVolumeAttachments(target.Namespace).Get(objName)
 			} else if cachedClient != nil {
-				err = cachedClient.Get(ctx, types.NamespacedName{Namespace: target.Namespace, Name: objName}, updateTarget)
+				originalObj = &azdiskv1beta2.AzVolumeAttachment{}
+				err = cachedClient.Get(ctx, types.NamespacedName{Namespace: target.Namespace, Name: objName}, originalObj)
 			} else {
-				updateTarget, err = azDiskClient.DiskV1beta2().AzVolumeAttachments(target.Namespace).Get(ctx, objName, metav1.GetOptions{})
-			}
-			if err == nil {
-				objForUpdate = updateTarget
-				copyForUpdate = updateTarget.DeepCopy()
+				originalObj, err = azDiskClient.DiskV1beta2().AzVolumeAttachments(target.Namespace).Get(ctx, objName, metav1.GetOptions{})
 			}
 		case *storagev1.VolumeAttachment:
-			updateTarget := &storagev1.VolumeAttachment{}
 			if cachedClient != nil {
-				err = cachedClient.Get(ctx, types.NamespacedName{Namespace: target.Namespace, Name: objName}, updateTarget)
+				originalObj = &storagev1.VolumeAttachment{}
+				err = cachedClient.Get(ctx, types.NamespacedName{Namespace: target.Namespace, Name: objName}, originalObj)
 			} else {
 				return status.Errorf(codes.Internal, "cannot update VolumeAttachment object if controller runtime client is not provided.")
-			}
-			if err == nil {
-				objForUpdate = updateTarget
-				copyForUpdate = updateTarget.DeepCopy()
 			}
 		default:
 			return status.Errorf(codes.Internal, "object (%v) not supported.", reflect.TypeOf(target))
@@ -1067,12 +1054,13 @@ func UpdateCRIWithRetry(ctx context.Context, informerFactory azdiskinformers.Sha
 			return err
 		}
 
+		copyForUpdate := originalObj.DeepCopyObject().(client.Object)
 		if err = updateFunc(copyForUpdate); err != nil {
 			return err
 		}
 
 		// if updateFunc doesn't change the object, don't bother making an update request
-		if reflect.DeepEqual(objForUpdate, copyForUpdate) {
+		if reflect.DeepEqual(originalObj, copyForUpdate) {
 			updatedObj = copyForUpdate
 			w.Logger().Info("Skip update. No update needed.")
 			return nil
@@ -1080,7 +1068,7 @@ func UpdateCRIWithRetry(ctx context.Context, informerFactory azdiskinformers.Sha
 
 		switch target := copyForUpdate.(type) {
 		case *azdiskv1beta2.AzVolume:
-			if (updateMode&UpdateCRIStatus) != 0 && !reflect.DeepEqual(objForUpdate.(*azdiskv1beta2.AzVolume).Status, target.Status) {
+			if (updateMode&UpdateCRIStatus) != 0 && !reflect.DeepEqual(originalObj.(*azdiskv1beta2.AzVolume).Status, target.Status) {
 				if updatedObj, err = azDiskClient.DiskV1beta2().AzVolumes(target.Namespace).UpdateStatus(ctx, target, metav1.UpdateOptions{}); err != nil {
 					return err
 				}
@@ -1089,7 +1077,7 @@ func UpdateCRIWithRetry(ctx context.Context, informerFactory azdiskinformers.Sha
 				updatedObj, err = azDiskClient.DiskV1beta2().AzVolumes(target.Namespace).Update(ctx, target, metav1.UpdateOptions{})
 			}
 		case *azdiskv1beta2.AzVolumeAttachment:
-			if (updateMode&UpdateCRIStatus) != 0 && !reflect.DeepEqual(objForUpdate.(*azdiskv1beta2.AzVolumeAttachment).Status, target.Status) {
+			if (updateMode&UpdateCRIStatus) != 0 && !reflect.DeepEqual(originalObj.(*azdiskv1beta2.AzVolumeAttachment).Status, target.Status) {
 				if updatedObj, err = azDiskClient.DiskV1beta2().AzVolumeAttachments(target.Namespace).UpdateStatus(ctx, target, metav1.UpdateOptions{}); err != nil {
 					return err
 				}
@@ -1099,7 +1087,7 @@ func UpdateCRIWithRetry(ctx context.Context, informerFactory azdiskinformers.Sha
 			}
 
 		case *storagev1.VolumeAttachment:
-			if (updateMode&UpdateCRIStatus) != 0 && !reflect.DeepEqual(objForUpdate.(*storagev1.VolumeAttachment).Status, target.Status) {
+			if (updateMode&UpdateCRIStatus) != 0 && !reflect.DeepEqual(originalObj.(*storagev1.VolumeAttachment).Status, target.Status) {
 				if err = cachedClient.Status().Update(ctx, target); err != nil {
 					return err
 				}
@@ -1110,7 +1098,19 @@ func UpdateCRIWithRetry(ctx context.Context, informerFactory azdiskinformers.Sha
 				updatedObj = target
 			}
 		}
-		return err
+
+		if err != nil {
+			// Return the raw error from the Update[Status] call here since the isRetriable check relies on this to determine
+			// whether or not to retry.
+			return err
+		}
+
+		if updatedObj == nil {
+			w.Logger().Info("No update applied - returning original object.")
+			updatedObj = originalObj.DeepCopyObject().(client.Object)
+		}
+
+		return nil
 	}
 
 	curRetry := 0

